Add support for "package extensions" to code loading (#47695)
JuliaLang/julia: doc/src/manual/code-loading.md

2. Packages in non-primary environments can end up using incompatible versions of their dependencies even if their own environments are entirely compatible. This can happen when one of their dependencies is shadowed by a version in an earlier environment in the stack (either by graph or path, or both).

Since the primary environment is typically the environment of a project you're working on, while environments later in the stack contain additional tools, this is the right trade-off: it's better to break your development tools but keep the project working. When such incompatibilities occur, you'll typically want to upgrade your dev tools to versions that are compatible with the main project.
### "Extension"s

> An "extension" is a module that is automatically loaded when a specified set of other packages (its "extension dependencies") are loaded in the current Julia session. The extension dependencies of an extension is a subset of those packages listed under the `[weakdeps]` section of a Project file. Extensions are defined under the `[extensions]` section in the project file:

Reviewer: ```suggestion
An "extension" is a module that is automatically loaded when a specified set of other packages (its "extension dependencies") are loaded in the current Julia session. The extension dependencies of an extension are a subset of those packages listed under the `[weakdeps]` section of a Project file. Extensions are defined under the `[extensions]` section in the project file:
```

These are defined by adding the following two sections to a package's `Project.toml` file:

```toml
name = "MyPackage"

[gluedeps]
GlueDep = "c9a23..." # uuid
OtherGlueDep = "862e..." # uuid

[gluepkgs]
GlueFoo = "GlueDep"
GlueBar = ["GlueDep", "OtherGlueDep"]
...
```

The keys under `gluepkgs` are the name of the glue packages.
They are loaded when all the packages on the right hand side (the glue dependencies) of the glue package are loaded.

> If a glue package only has one glue dependency the lit of glue dependencies can be written as just a string for breviety.

Reviewer 2: ```suggestion
If a glue package only has one glue dependency, the list of glue dependencies can be written as just a string for brevity.
```

2. Packages in non-primary environments can end up using incompatible versions of their dependencies even if their own environments are entirely compatible. This can happen when one of their dependencies is shadowed by a version in an earlier environment in the stack (either by graph or path, or both).

Since the primary environment is typically the environment of a project you're working on, while environments later in the stack contain additional tools, this is the right trade-off: it's better to break your development tools but keep the project working. When such incompatibilities occur, you'll typically want to upgrade your dev tools to versions that are compatible with the main project.

> ### "Glue" packages and dependencies

Reviewer 3: Obviously there are more changes to like this, but just to mark it as I think this is where the discussion settled
```suggestion
### "Weak" dependencies and  "Glue" packages
```

A "glue package" is a module that is automatically loaded when a specified set of other packages (called "glue dependencies") are loaded in the current Julia session.
These are defined by adding the following two sections to a package's `Project.toml` file:

```toml
name = "MyPackage"

> [gluedeps]
> GlueDep = "c9a23..." # uuid
> OtherGlueDep = "862e..." # uuid

Reviewer 3: ```suggestion
[weakdeps]
WeakDep = "c9a23..." # uuid
OtherWeakDep = "862e..." # uuid
```

[gluepkgs]

> GlueFoo = "GlueDep"
> GlueBar = ["GlueDep", "OtherGlueDep"]

Reviewer 3: I think setting up a convention of *Glue would be better here. i.e. `FooGlue` reads like _the glue for Foo_
```suggestion
WeakDepGlue = "WeakDep"
FooGlue = ["WeakDep", "OtherWeakDep"]
```

...
```

The keys under `gluepkgs` are the name of the glue packages.

> They are loaded when all the packages on the right hand side (the glue dependencies) of the glue package are loaded.

Reviewer 3: ```suggestion
They are loaded when all the packages on the right hand side (the weak dependencies) of the glue package are loaded.
```
